src/nn/transformer.rs

```diff
@@ -0,0 +1,1097 @@
+    Assert<{ B * H * S * (V / H) == B * S * V }>: ConstTrue,
+{
+    type Output = Tensor3D<B, S, M>;
+
+    fn forward(&self, input: Tensor3D<B, S, M>) -> Self::Output {
+        let queries = self.w_q.forward(input.duplicate());
+        let keys = self.w_k.forward(input.duplicate());
+        let values = self.w_v.forward(input);
+
+        let keys: Tensor4D<B, H, S, { K / H }> = keys.reshape();
+        let queries: Tensor4D<B, H, S, { K / H }> = queries.reshape();
+        let values: Tensor4D<B, H, S, { V / H }> = values.reshape();
+
+        // Get weights
+        let token_weights = batch_4d_matmul_transpose(queries, &keys) / (M as f32);
+
+        // Softmax on last dimension
+        let token_weights = softmax(token_weights);
+
+        // Get new tokens
+        let tokens: Tensor4D<B, H, S, { V / H }> = batch_4d_matmul(token_weights, &values);
+        let tokens: Tensor3D<B, S, V> = tokens.reshape();
+        self.w_o.forward(tokens)
+    }
+}
+
+/// Batched Encoder-Decoder style self attention where one set of tensors is used for values and keys, and another is used for queries
+impl<
+        const B: usize,
+        const M: usize,
+        const N: usize,
+        const K: usize,
+        const V: usize,
+        const S1: usize,
+        const S2: usize,
+        const H: usize,
+    > Module<(Tensor3D<B, S1, M>, Tensor3D<B, S2, N>)> for MultiHeadAttention<M, N, K, V, H>
+where
+    Assert<{ B * S2 * K == B * H * S2 * (K / H) }>: ConstTrue,
+    Assert<{ B * S1 * K == B * H * S1 * (K / H) }>: ConstTrue,
+    Assert<{ B * S1 * V == B * H * S1 * (V / H) }>: ConstTrue,
+    Assert<{ B * S2 * H * { V / H } == B * S2 * V }>: ConstTrue,
+    Assert<{ B * H * S2 * (V / H) == B * S2 * V }>: ConstTrue,
+{
+    type Output = Tensor3D<B, S2, M>;
+
+    fn forward(&self, (from_enc, input): (Tensor3D<B, S1, M>, Tensor3D<B, S2, N>)) -> Self::Output {
+        let queries = self.w_q.forward(input);
+        let keys = self.w_k.forward(from_enc.duplicate());
+        let values = self.w_v.forward(from_enc);
+
+        let keys: Tensor4D<B, H, S1, { K / H }> = keys.reshape();
+        let queries: Tensor4D<B, H, S2, { K / H }> = queries.reshape();
+        let values: Tensor4D<B, H, S1, { V / H }> = values.reshape();
+
+        // Get weights
+        let token_weights = batch_4d_matmul_transpose(queries, &keys) / (M as f32);
+
+        // Softmax on last dimension
+        let token_weights = softmax(token_weights);
+
+        // Get new tokens
+        let tokens: Tensor3D<B, S2, V> = batch_4d_matmul(token_weights, &values).reshape();
+        self.w_o.forward(tokens)
+    }
+}
+
+/// A single transformer block containing self attention, feed forward and layer norms
+#[derive(Debug, Clone, Default)]
+pub struct TransformerBlock<const M: usize, const I: usize, const K: usize, const H: usize> {
+    attn: MultiHeadAttention<M, M, K, M, H>,
+    norm1: LayerNorm1D<M>,
+    norm2: LayerNorm1D<M>,
+    ff: (Linear<M, I>, ReLU, Linear<I, M>),
+}
```

Comment:
Could this be implemented as
```rust
type TransformerBlock<....> = (
    Residual<MultiHeadAttention<...>>,
    LayerNorm1D<M>,
    Residual<(Linear, ReLU, Linear)>,
)
```

I was envisioning using `Residual` a bunch at this level.

Reply:
Awesome!